netbox/extras/filtersets.py

```diff
@@ -434,8 +435,9 @@ class JobResultFilterSet(BaseFilterSet):
         method='search',
         label='Search',
     )
-    created = django_filters.DateTimeFilter()
-    completed = django_filters.DateTimeFilter()
+    created = django_filters.DateTimeFromToRangeFilter()
+    completed = django_filters.DateTimeFromToRangeFilter()
```

Comment:
Is this a breaking change, or is the behavior of the original filter retained?

Reply:
I looked into this now, and I'm not sure what is the right solution here and I'm frankly a little confused.

Current behavior (develop branch) is that there's a created and completed DateTimeFilter.

https://github.com/netbox-community/netbox/blob/96c469641796b2b12ce643363d41d823670b68db/netbox/extras/filtersets.py#L437-L438

In practice this means that you can filter on the exact datetime value, so like `created=2022-10-21T09:23:57.593918`. Neither gte or lte works here. I think for those to work the filterset had to look like this?

https://github.com/netbox-community/netbox/blob/96c469641796b2b12ce643363d41d823670b68db/netbox/users/filtersets.py#L79-L87

Now with my change to DateTimeFromToRangeFilter created_after, created_before, completed_after, completed_before are added. So there's an issue here that the base created/completed actually doesn't exist anymore. So how should this be added to the filterset to make the swagger definition correct?

If we look at another place where DateTimeFromToRange is used, it shows the same issue:

https://github.com/netbox-community/netbox/blob/96c469641796b2b12ce643363d41d823670b68db/netbox/extras/filtersets.py#L399

The `time` filter that shows up in swagger doesn't actually work at all, but `time_before` and `time_after` does.

I'm really not sure what to do here, it needs to be DateTimeFromToRangeFilter for the filter form to work like for example the object change filter form. But the filters need to be different for the API for the swagger docs to match. I don't think it makes sense to have a filter that only matches on exact date, but __gte and __lte works some places like the TokenFilterSet I linked above, should we use a normal DateTimeFilter with those instead?

Reply:
Agreed, this is frustrating.

> The time filter that shows up in swagger doesn't actually work at all, but time_before and time_after does.

AFAICT this is because DateTimeFromToRangeFilter doesn't actually register the before/after lookups on the filterset, so they can't be inspected. `get_filters()` returns a single filter for each:

```
created <django_filters.filters.DateTimeFromToRangeFilter object at 0x7fcf591b2c40>
completed <django_filters.filters.DateTimeFromToRangeFilter object at 0x7fcf591b2c70>
```

It's not ideal, but the simplest approach until we can come up with something more elegant would be to add the before & after lookups explicitly, like under TokenFilterSet. However I do think we should stick with the before/after nomenclature as they make more sense in the context of time than lte/gte. Something like this:

```
created = django_filters.DateTimeFilter()
created_before = django_filters.DateTimeFilter(
    field_name='created',
    lookup_expr='lte'
)
created_after = django_filters.DateTimeFilter(
    field_name='created',
    lookup_expr='gte'
)
```

It's probably worth auditing how we handle datetime filtering in NetBox overall, but this approach would let us knock out this PR without breaking either of the existing filters. What do you think?

Reply:
Looks perfect, thanks!

> Do note that it is now inconsistent with the ObjectChangeFilterSet and FilterForm

Yep, and I think that's fine for now. Ultimately (possibly in v3.4) I expect we'll bring them in line with what you have here, to be more consistent with other filter lookups in general.